### DISCUSSION TITLE
Possible dependency incompatibility when uploading files

### DISCUSSION THREAD

**Discussion by g1sbi** [pocketbase/pocketbase#2768] [Q&A] [answered]
I'm uploading images in a node environment, but get this error. I recently had to update all my dependencies to fix a bug, so it could be caused by that.

```ts
/home/gisbi/code/thock/thock-scraper/node_modules/delayed-stream/lib/delayed_stream.js:33
  source.on('error', function() {});
         ^
TypeError: source.on is not a function
    at Function.DelayedStream.create (/home/gisbi/code/thock/thock-scraper/node_modules/delayed-stream/lib/delayed_stream.js:33:10)
    at FormData.CombinedStream.append (/home/gisbi/code/thock/thock-scraper/node_modules/combined-stream/lib/combined_stream.js:45:37)
    at FormData.append (/home/gisbi/code/thock/thock-scraper/node_modules/form-data/lib/form_data.js:75:3)
    at /home/gisbi/code/thock/thock-scraper/geekhack-scraper/saveToPocketbase.ts:63:16
    at Generator.next (<anonymous>)
    at /home/gisbi/code/thock/thock-scraper/geekhack-scraper/saveToPocketbase.ts:8:71
    at new Promise (<anonymous>)
    at __awaiter (/home/gisbi/code/thock/thock-scraper/geekhack-scraper/saveToPocketbase.ts:4:12)
    at /home/gisbi/code/thock/thock-scraper/geekhack-scraper/saveToPocketbase.ts:61:39
    at Array.map (<anonymous>)
```

this is the code:
```ts
async function uploadImages(images: Image[], thread_id: string) {
  const imageIds: string[] = [];
  await Promise.all(
    images.map(async (image: Image) => {
      const formData = new FormData();
      formData.append(
        "documents",
        new Blob([
          fs.readFileSync("images/" + thread_id + "/" + image.name),
        ]),
        image.name
      );

      const data = {
        thread_id: thread_id,
        image: formData,
        sort_order: image.sort_order,
      };
      const createdRecord = await pb.collection("images").create(data);

      imageIds.push(createdRecord.id);
    })
  );
  return imageIds;
}
```

**Comment by ganigeorgiev** [comment]
You have to use either `FormData` for sending `multipart/form-data` request or the JS object for sending `application/json`.
File upload is supported only via `multipart/form-data`. You could try:

```js
const formData = new FormData();

formData.append(
  "documents",
  new Blob([fs.readFileSync("images/" + thread_id + "/" + image.name)]),
  image.name
);

formData.append("thread_id", thread_id)

formData.append("sort_order", image.sort_order)

const createdRecord = await pb.collection("images").create(formData);
```

Option to pass a `File`/`Blob` using the JS object syntax may be added eventually in the future since several other users requested it (_you can subscribe to https://github.com/pocketbase/js-sdk/issues/209_).

**Reply by g1sbi** [reply]
Same error, unfortunately

**Reply by ganigeorgiev** [reply]
I'm not sure what could be the cause of the error. Passing a `FormData` object with `fs.readFileSync` Blob works fine for me with Node v18.10.

Note that the JS SDK will use any global `fetch` and `FormData` implementation that is available in your environment, so if you use a polyfill you can try to upgrade to Node 18+ since those apis are natively supported and remove the polyfills.

**Reply by ganigeorgiev** [reply]
_Also keep in mind that the above will execute concurrently and [autocancellation](https://github.com/pocketbase/js-sdk#auto-cancellation) may get triggered. To disable it globally you can use `pb.autoCancellation(false)` or disable it per request by passing `create(data, { '$autoCancel': false})`._

**Reply by g1sbi** [reply]
I retried using node 18.10.0  and autocancellation disabled globally, but no change.

**Reply by ganigeorgiev** [reply]
I'm not sure how to help. The JS SDK doesn't have any external dependencies. As mentioned earlier, we use whatever global `fetch` and `FormData` are available in the environment.

If you are still not able to resolve the issue, please share a minimal reproducible repo or at least provide some reproducible steps.

**Reply by g1sbi** [reply]
https://github.com/g1sbi/pocketbase-fileupload-testing

here's a repo with the same error and same server setup.

**Reply by ganigeorgiev** [reply, marked as answer]
I suspect that the `form-data` polyfill is conflicting with the native `fetch`. Try to remove it and use the native `FormData` available with Node.js 18+.

**Reply by g1sbi** [reply]
That was it, thanks. However, after the upload completes successfully, the file's field is still empty. 
Console logging the `formData` variable gives this 
```ts
FormData {
  [Symbol(state)]: [
    { name: 'documents', value: [File] },
    { name: 'thread_id', value: '110967' },
    { name: 'sort_order', value: '55' }
  ]
}
```

which seems correct. How do I debug this further?

**Reply by ganigeorgiev** [reply]
Make sure that you use the correct file field name.

You can also enable the debug mode with `--debug` and check the console output for some error/warning during the upload.

**Reply by g1sbi** [reply]
Never mind, fixed. I didn't realize that I should've used the actual parameter name instead of "documents", now it works. Thanks again!
